### PR TITLE
Feature/python annotations of py4j types

### PR DESCRIPTION
**Pull request by migwellian**
I ran `stubgen` on `py4j` and changed most of our `Any` type annotations to Py4J types.

UPDATE: I also moved `VertexOps` into `Vertex` because the type annotations caused cyclic dependencies when they were separate files. It makes more sense logically too.

This PR is not as big as it looks, as a lot of the additions are the stub files for dependencies, which are mostly auto-generated.

### CODE REVIEW

**Comment by icantsayrural** [comment]
Two things:
- Will need to manually verify that the types are recognised
- Type the rest of the tests as well

**Comment by migwellian** [comment]
done and done

File: keanu-python/keanu/vertex/base.py [hunk 1]
```diff
@@ -1,24 +1,33 @@
...
-from keanu.cast import cast_tensor_arg_to_double, cast_tensor_arg_to_integer, cast_tensor_arg_to_bool
 
 k = KeanuContext()
 
+vertex_operation_param_types = Union['Vertex', tensor_arg_types]
+vertex_constructor_param_types = Union['Vertex', tensor_arg_types, wrapped_java_types]
+
 
-class Vertex(JavaObjectWrapper, VertexOps):
+class Vertex(JavaObjectWrapper, SupportsRound['Vertex']):
```
[the remark this answers is not in the document]
Comment (migwellian) [reply keanu-python/keanu/vertex/base.py]: it's a nifty MyPy class to make typing work with the built-in `round` function.

Unfortunately it's slightly broken - see https://stackoverflow.com/questions/53481887/looking-for-a-working-example-of-supportsround

File: keanu-python/tests/test_model.py [hunk 2]
```diff
@@ -21,9 +21,9 @@ def test_to_bayes_net():
     assert m.gaussian.get_id() in net_vertex_ids
 
 
-def test_to_bayes_net_excludes_non_vertices():
+def test_to_bayes_net_excludes_non_vertices() -> None:
     with Model() as m:
-        m.not_a_vertex = 1
+        m.not_a_vertex = 1  # type: ignore # because Model#__set_attr__ expects Vertex
```
[the remark this answers is not in the document]
Comment (migwellian) [reply keanu-python/tests/test_model.py]: Yes.

I thought you deliberately forced it to take `Vertex`?

File: keanu-python/examples/coal_mining.py [hunk 3]
```diff
@@ -19,11 +19,11 @@ def model(self):
...
 
-            m.years = np.array(self._data.index)
+            m.years = np.array(self._data.index)  # type: ignore # because Model#__set_attr__ expects Vertex
             m.rates = DoubleIf(m.switchpoint > m.years, m.early_rate, m.late_rate)
             m.disasters = Poisson(m.rates)
 
         return m
 
-    def training_data(self):
+    def training_data(self) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (migwellian) [reply keanu-python/examples/coal_mining.py]: have replaces `np.ndarray` with `numpy_types` throughout - thanks.

File: keanu-python/tests/test_net.py [hunk 4]
```diff
@@ -24,7 +24,7 @@ def test_construct_bayes_net():
                           ("get_observed_vertices", False, True, True, True),
                           ("get_continuous_latent_vertices", True, False, True, False),
                           ("get_discrete_latent_vertices", True, False, False, True)])
-def test_can_get_vertices_from_bayes_net(get_method, latent, observed, continuous, discrete):
+def test_can_get_vertices_from_bayes_net(get_method, latent, observed, continuous, discrete) -> None:
```
[the remark this answers is not in the document]
Comment (migwellian) [reply keanu-python/tests/test_net.py]: Not on purpose, although I think this is the only one I missed?

File: keanu-python/tests/test_vertex.py [hunk 5]
```diff
@@ -36,54 +40,88 @@ def assert_vertex_value_equals_pandas(vertex, expected_type, pandas):
...
     with pytest.raises(ValueError) as excinfo:
-        Vertex(jvm_view.GaussianVertex, GenericExampleClass(), GenericExampleClass())
+        Vertex(  # type: ignore # this is expected to fail mypy
+            jvm_view.GaussianVertex, GenericExampleClass(), GenericExampleClass())
 
     assert str(excinfo.value) == "Can't parse generic argument. Was given {}".format(GenericExampleClass)
 
 
-def test_int_vertex_value_is_a_numpy_array():
+def test_you_can_set_and_get_a_value(jvm_view) -> None:
```
[the remark this answers is not in the document]
Comment (migwellian) [reply keanu-python/tests/test_vertex.py]: ok - are you suggesting I just delete this test?

[the remark this answers is not in the document]
Comment (icantsayrural) [reply keanu-python/tests/test_vertex.py]: Yeah

File: keanu-python/tests/test_model.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (icantsayrural) [reply keanu-python/tests/test_model.py]: No, especially if it makes users add `# type: ignore` whenever they add non-vertices to the bayes net

File: keanu-python/keanu/vertex/base.py [hunk 6]
```diff
@@ -1,24 +1,33 @@
...
-    def __init__(self, val: Any, *args: Union[vertex_param_types, shape_types]) -> None:
+    def __init__(self, val_or_ctor: Union[JavaMember, JavaObject],
+                 *args: Union[vertex_constructor_param_types, shape_types]) -> None:
+        val: JavaObject
         if args:
-            ctor = val
+            ctor = val_or_ctor
             val = ctor(*(Vertex.__parse_args(args)))
+        else:
+            val = cast(JavaObject, val_or_ctor)
```
[the remark this answers is not in the document]
Comment (icantsayrural) [reply keanu-python/keanu/vertex/base.py]: Or just rename `Vertex.cast`

File: keanu-python/tests/test_net.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (icantsayrural) [reply keanu-python/tests/test_net.py]: Some test and helper methods in `test_vertices` are not, I don't think

File: keanu-python/keanu/vertex/base.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (migwellian) [reply keanu-python/keanu/vertex/base.py]: Wait, I'm confused now - `Vertex.cast` doesn't do anything!

    def cast(self, v: tensor_arg_types) -> tensor_arg_types:
        return v

File: keanu-python/tests/test_net.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (migwellian) [reply keanu-python/tests/test_net.py]: Oh yes, I see now. I was only looking at those annotated with `@pytest.mark.parametrize`

File: keanu-python/tests/test_vertex.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (migwellian) [reply keanu-python/tests/test_vertex.py]: done

File: keanu-python/tests/test_model.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (migwellian) [reply keanu-python/tests/test_model.py]: done

File: keanu-python/keanu/vertex/base.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (icantsayrural) [reply keanu-python/keanu/vertex/base.py]: It's overriden by `Double`, `Integer` and `Bool`.

`Vertex` should still be able to be instantiated (i.e. isn't an ABC so need to implement `cast`), cause sometimes we can't infer the type

**Pull request merged**